cmd/lncli/commands.go

```diff
@@ -728,28 +756,224 @@ func closeChannel(ctx *cli.Context) error {
+
+		// We can safely exit if the user did not confirm.
+		if !confirmed {
+			return nil
+		}
+
+		channelsToClose = openChannels.Channels
+	default:
+		// Go through the list of open channels and determine which
+		// should be added to the closing list.
+		for _, channel := range openChannels.Channels {
+			// If the channel is inactive, we'll attempt to
+			// unilaterally close the channel, so we should prompt
+			// the user for confirmation beforehand.
+			if !channel.GetActive() {
+				msg := fmt.Sprintf("Unilaterally close channel "+
+					"with node %s and channel point %s? "+
+					"The closing transaction will need %d "+
+					"confirmations before the funds can be "+
+					"spent. (yes/no): ", channel.RemotePubkey,
+					channel.ChannelPoint, channel.CsvDelay)
+
+				confirmed := promptForConfirmation(msg)
+
+				if confirmed {
+					channelsToClose = append(
+						channelsToClose, channel,
+					)
+				}
+			} else if !ctx.Bool("inactive_only") {
+				// Otherwise, we'll only add active channels if
+				// we were not requested to close inactive
+				// channels only.
+				channelsToClose = append(
+					channelsToClose, channel,
+				)
+			}
+		}
+	}
+
+	// result defines the result of closing a channel. The closing
+	// transaction ID is populated if a channel is successfully closed.
+	// Otherwise, the error that prevented closing the channel is populated.
+	type result struct {
+		RemotePubKey string `json:"remote_pub_key"`
+		ChannelPoint string `json:"channel_point"`
+		ClosingTxid  string `json:"closing_txid"`
+		FailErr      string `json:"error"`
+	}
+
+	// Launch each channel closure in a goroutine in order to execute them
+	// in parallel. Once they're all executed, we will print the results as
+	// they come.
+	resultChan := make(chan result, len(channelsToClose))
+	for _, channel := range channelsToClose {
+		go func(channel *lnrpc.ActiveChannel) {
+			res := result{}
+			res.RemotePubKey = channel.RemotePubkey
+			res.ChannelPoint = channel.ChannelPoint
+			defer func() {
+				resultChan <- res
+			}()
+
+			// Parse the channel point in order to create the close
+			// channel request.
+			s := strings.Split(res.ChannelPoint, ":")
+			if len(s) != 2 {
+				res.FailErr = "expected channel point with " +
+					"format txid:index"
+				return
+			}
+			index, err := strconv.ParseUint(s[1], 10, 32)
 			if err != nil {
-				return err
+				res.FailErr = fmt.Sprintf("unable to parse "+
+					"channel point output index: %v", err)
+				return
 			}
 
-			printJSON(struct {
-				ClosingTXID string `json:"closing_txid"`
-			}{
-				ClosingTXID: txid.String(),
-			})
+			req := &lnrpc.CloseChannelRequest{
+				ChannelPoint: &lnrpc.ChannelPoint{
+					FundingTxid: &lnrpc.ChannelPoint_FundingTxidStr{
+						FundingTxidStr: s[0],
+					},
+					OutputIndex: uint32(index),
+				},
+				Force: !channel.GetActive(),
+			}
+
+			txidChan := make(chan string, 1)
+			err = executeChannelClose(ctx, req, txidChan)
```

Comment:
One potential problem with passing in the context here (haven't tested it): what if you call `closeallchannels` with `block=true`? Would maybe be safer to let `executeChannelClose` take the `client` and a `boolean` and in this case just setting it false.

Reply:
Nice catch!